### ISSUE TITLE
Converting Compiler Explorer to C++ - Fork Announcement

### ISSUE THREAD

**Issue by BoredBored** [mattgodbolt/gcc-explorer#197]
So before you mark this off as spam*, please just read the rest of what I'm going to say

Overview
-----------
Ok, so I'll get so the point - what I am doing is converting Compiler Explorer from JavaScript to C++. 
The following are reasons why I am doing this:

- I am building an IDE in C++ and one of the features I want to cram in there is converting C++ code into asm so I can get _just a little bit more_ performance

- In case anyone wants a faster, out of the box version version of CE to download

- To create a version of CE that will be less prone to bugs and errors and be more upgradeable & organized

Goals:
-------

#### Short-Term:

- Start and finish converting the current version of CE to C++ (About 6 or 5 weeks)
- Make the source code files more organized

### Long-Term:

- Make a out of the box version of CE that's beginner friendly
- Possibly add support for Python
- Finish IDE
- When the IDE is finished, work on adding support for converting C++ code to Python and vice versa  using deep learning

Anyway I just wanted to say that I'm making a different version of CE

*I had no idea where to post this where people would see it so I just put it here - sorry

**Comment by mattgodbolt** [comment]
Hi!

Thanks so much for posting this. It's an interesting idea, and I have some quick thoughts, though I'd be happy to discuss further:

* Compiler Explorer doesn't really turn C++ etc into asm. It runs a compiler (which does this very hard work), then does a little bit of reformatting, and presentation in a browser.
* I'm not sure what speed issues you may be running into, but the transformations typically don't take very long and are usually dwarfed by the compilation time itself. If you have any counter examples I'd love to see them as I'd like to improve the JS version too.
* You say that making a C++ version will be "less prone to bugs and errors and more upgradeable". I've been programming C++ for nearly 25 years and that's never been my experience. Writing is C++ is one of the harder things to do which is why one rarely writes C++ unless there's a very good reason. Have you found Compiler Explorer particularly prone to bugs?
* I agree the source isn't very readable and isn't well organized. I believe that's orthogonal to which language it's written in: the haphazard code here is the result of four years of organic growth, plus my learning of Javascript!
* The majority of the complexity in CE is in curating and maintaining the many compilers in the backend, and making the frontend code deal with them (somewhat!) transparently.
* Converting C++ code to Python and vice versa using deep learning sounds extraordinarily out of scope for something like CE!

If there's anything in this brain-dump of thoughts that doesn't make sense, please feel free to follow-up..I'm not quite sure what to make of this. Of course I wish you well in your endeavours and if I can help at all, do let me know.

**Comment by BoredBored** [comment]
Dam you type fast! Anyway -

>Compiler Explorer doesn't really turn C++ etc into asm. It runs a compiler (which does this very hard work), then does a little bit of reformatting, and presentation in a browser.

Oh, my mistake, but It's still good for my purposes.

> I'm not sure what speed issues you may be running into, but the transformations typically don't take very long and are usually dwarfed by the compilation time itself. If you have any counter examples I'd love to see them as I'd like to improve the JS version too.

If you type code in [CE][1] you will get a kind of flash where the page is white then it updates*. Some people might say "meh, It doesn't matter", but I won't. The update time needs to go down to where a person will not be able to tell that there is a lag at all.

*If the lag is from the compilers you got me

>You say that making a C++ version will be "less prone to bugs and errors and more upgradeable". I've been programming C++ for nearly 25 years and that's never been my experience.

What I mean is that if I create a C++ version now based on the original version of CE that I can make it more organized, thus if its easier to find and edit later instead of a jumbled mess fewer bug are able to happen. I'm not saying that you don't know how to program (in fact you probably know more then me when it comes to experience) but bugs do happen - and even if they're easily fixable, the fact is you can make code better with organization.

>Writing is C++ is one of the harder things to do which is why one rarely writes C++ unless there's a very good reason. 

I just love programming in C/C++. Let's me have more control over how my programs operate then,  you could say, most programming languages .

>Have you found Compiler Explorer particularly prone to bugs?
Nope. Not really.

>I agree the source isn't very readable and isn't well organized. I believe that's orthogonal to which language it's written in: the haphazard code here is the result of four years of organic growth, plus my learning of Javascript!

So you agree, it's time for a change?

>The majority of the complexity in CE is in curating and maintaining the many compilers in the backend, and making the frontend code deal with them (somewhat!) transparently.

How long did it take you to make CE to date?

>Converting C++ code to Python and vice versa using deep learning sounds extraordinarily out of scope for something like CE!

Whys that? Even though it's called compiler explorer I think I'd be great if there could be a python to C++ source code converter out there*

*I did a google search "c++ code converter" and found that they either didn't exist or you had to pay

[1]:https://godbolt.org/

**Comment by BoredBored** [comment]
So, the link to the Github page is [https://github.com/BoredBored/Compiler-Explorer-IDE][1]

[1]:https://github.com/BoredBored/Compiler-Explorer-IDE

**Comment by mattgodbolt** [comment]
A few replies:

* The "white flash" is the delay I impose (on purpose) before running the colourisation code. That delay is there to allow multiple compilers to do their thing. Multiple compiles run in parallel, but the colourisation is a global process that has to re-run every time a compiler completes. Rather than have multiple updates (where the colours change every few milliseconds as multiple compilers finish their work), I delay a bit. I'm lazily using a time-based delay, instead of tracking all the possible multiple compilation results to complete.  Whichelver way you cut it, there would still be a delay as the compilation is remote. Any ideas how to make this look more appealing very welcomed!
* Re "time for a change" - not to a completely new language, no. I'd happily accept pull requests to tidy up the code here (and indeed have a few myself nearly ready to go in).
* Re: "How long to date" - around four years, on and off.
* Re: "Converting C++ code to Python and vice versa" - I wish you the best of luck with that; but I suspect you may find some interesting (and in my view, unsolvable) problems along the way. Please don't let an old man potentially jaded with the wrong kind of experience put you off trying though!

**Comment by mattgodbolt** [comment]
Happy to discuss further (here, or over email), but closing as not really relevant to Compiler Explorer.
